File: src/server/chain/prices.ts
import { USDG, WETH } from './abi';
import type { GeckoClient } from './gecko';

const HOUR_S = 3600;
const NEAR_S = 6 * HOUR_S; // thin quote markets skip hours; a candle this close is used as is
const FAR_S = 48 * HOUR_S; // beyond this there is no honest price for that hour
// The deepest USDG/WETH pool on the chain prices WETH.
const REFERENCE_POOLS: Record<string, string> = { [WETH]: '0x52e65b17fb6e5ba00ed806f37afcd2daa50271ca' };
const STABLES = new Set([USDG]);

/** Hourly USD prices of the quote assets tokens launch against: WETH, USDG and tokenized stocks. */
export class QuotePrices {
  private readonly closes = new Map<string, Map<number, number>>();
  private readonly covered = new Map<string, [number, number][]>();
  private readonly pools = new Map<string, Promise<string | null>>();

  constructor(
    private readonly gecko: Pick<GeckoClient, 'hourlyCloses' | 'topPool'>,
    private readonly nowSeconds: () => number,
  ) {}

  async usdAt(quote: string, seconds: number): Promise<number> {
    const q = quote.toLowerCase();
    if (STABLES.has(q)) return 1;
    const hour = Math.floor(seconds / HOUR_S) * HOUR_S;
    const series = this.closes.get(q) ?? new Map<number, number>();
    this.closes.set(q, series);

    let price = nearest(series, hour, NEAR_S);
    const covered = this.covered.get(q) ?? [];
    if (price == null && !covered.some(([from, to]) => hour >= from && hour <= to)) {
      const pool = await this.poolFor(q);
      if (!pool) throw new Error(`no USD market for quote ${q}`);
      const until = Math.min(this.nowSeconds(), hour + 500 * HOUR_S);
      for (const [t, close] of await this.gecko.hourlyCloses(pool, q, until, 1000)) if (close > 0) series.set(t, close);
      covered.push([until - 999 * HOUR_S, until - HOUR_S]); // the newest hour is still open and fetched again later
      this.covered.set(q, covered);
      price = nearest(series, hour, NEAR_S);
    }
    price ??= nearest(series, hour, FAR_S);
    if (price == null) throw new Error(`no USD price for ${q} near ${new Date(hour * 1000).toISOString()}`);
    return price;
  }

  private poolFor(quote: string) {
    let pool = this.pools.get(quote);
    if (!pool) {
      pool = REFERENCE_POOLS[quote] ? Promise.resolve(REFERENCE_POOLS[quote]) : this.gecko.topPool(quote);
      pool.catch(() => this.pools.delete(quote));
      this.pools.set(quote, pool);
    }
    return pool;
  }
}

function nearest(series: Map<number, number>, hour: number, within: number): number | null {
  for (let d = 0; d <= within; d += HOUR_S) {
    const price = series.get(hour - d) ?? series.get(hour + d);
    if (price != null) return price;
  }
  return null;
}

File: backend/wassily/chain/prices.py
"""Hourly USD prices of the quote assets tokens launch against: WETH, USDG and tokenized stocks."""

from __future__ import annotations

import time
from collections.abc import Callable
from typing import Protocol

from wassily.chain.abi import USDG, WETH
from wassily.utils.timeutil import iso

HOUR_S = 3600
NEAR_S = 6 * HOUR_S  # thin quote markets skip hours; a candle this close is used as is
FAR_S = 48 * HOUR_S  # beyond this there is no honest price for that hour

# The deepest USDG/WETH pool on the chain prices WETH.
REFERENCE_POOLS = {WETH: "0x52e65b17fb6e5ba00ed806f37afcd2daa50271ca"}
STABLES = frozenset({USDG})


class CandleSource(Protocol):
    async def hourly_closes(self, pool: str, token: str, before_sec: int, limit: int = 1000) -> list[tuple[int, float]]:
        ...

    async def top_pool(self, token: str) -> str | None:
        ...


class NoUsdPrice(LookupError):
    pass


def _nearest(series: dict[int, float], hour: int, within: int) -> float | None:
    for d in range(0, within + 1, HOUR_S):
        price = series.get(hour - d)
        if price is None:
            price = series.get(hour + d)
        if price is not None:
            return price
    return None


class QuotePrices:
    def __init__(self, gecko: CandleSource, now_seconds: Callable[[], float] = time.time) -> None:
        self._gecko = gecko
        self._now = now_seconds
        self._closes: dict[str, dict[int, float]] = {}
        self._covered: dict[str, list[tuple[int, int]]] = {}
        self._pools: dict[str, str | None] = {}

    async def usd_at(self, quote: str, seconds: float) -> float:
        q = quote.lower()
        if q in STABLES:
            return 1.0
        hour = int(seconds // HOUR_S) * HOUR_S
        series = self._closes.setdefault(q, {})

        price = _nearest(series, hour, NEAR_S)
        covered = self._covered.setdefault(q, [])
        if price is None and not any(lo <= hour <= hi for lo, hi in covered):
            pool = await self._pool_for(q)
            if not pool:
                raise NoUsdPrice(f"no USD market for quote {q}")
            until = int(min(self._now(), hour + 500 * HOUR_S))
            for t, close in await self._gecko.hourly_closes(pool, q, until, 1000):
                if close > 0:
                    series[t] = close
            covered.append((until - 999 * HOUR_S, until - HOUR_S))  # the newest hour is still open
            price = _nearest(series, hour, NEAR_S)
        if price is None:
            price = _nearest(series, hour, FAR_S)
        if price is None:
            raise NoUsdPrice(f"no USD price for {q} near {iso(hour * 1000)}")
        return price

    async def _pool_for(self, quote: str) -> str | None:
        if quote in self._pools:
            return self._pools[quote]
        pool = REFERENCE_POOLS.get(quote) or await self._gecko.top_pool(quote)
        self._pools[quote] = pool  # a failed lookup raises before this line and is retried next time
        return pool
